lib/model/compose.dart

```diff
@@ -0,0 +1,112 @@
+import 'dart:math';
+
+//
+// Put functions for nontrivial message-content generation in this file.
+//
+// If it's complicated enough to need tests, it should go in here.
+//
+
+// https://spec.commonmark.org/0.30/#fenced-code-blocks
+final RegExp _openingBacktickFenceRegex = (() {
+  // Allow "up to three spaces of indentation".
+  // As of Zulip Server 7.0, we don't comply with that detail:
+  //   https://chat.zulip.org/#narrow/stream/6-frontend/topic/quote-and-reply.20fence.20length/near/1588273
+  // and that's a bug.
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest opening backtick fence that's longer than any in [content].
+///
+/// Expressed as a number of backticks.
+///
+/// Use this for quote-and-reply or anything else that requires wrapping
+/// Markdown in a backtick fence.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedOpeningBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length);
```

Comment:
`match[1]!.length + 1`?

Reply:
Eep, yep, thanks for the catch! An example of why we'll want this code to have tests before we merge it. 😅